src/predicate_lang/predicate_lang.ml

```diff
@@ -8,62 +8,75 @@ type 'a t =
   | Union of 'a t list
   | Inter of 'a t list
 
+let element a = Element a
+
+let of_list xs = Union (List.rev_map ~f:element xs)
+
+let standard = Standard
+
 let diff a = function
   | Union [] -> a
   | Inter [] -> Union []
   | b -> Inter [ a; Compl b ]
 
-let inter a = Inter a
-
 let compl a = Compl a
 
-let union a = Union a
+let true_ = Inter []
 
-let not_union a = compl (union a)
+let false_ = Union []
 
-let any = not_union []
+let union = function
+  | [] -> false_
+  | [ x ] -> x
+  | _ :: _ :: _ as xs -> Union xs
 
-let empty = Union []
+let inter = function
+  | [] -> true_
+  | [ x ] -> x
+  | _ :: _ :: _ as xs -> Inter xs
 
-let rec decode_one f =
-  let open Decoder in
-  let bool_ops () =
-    sum
-      [ ("or", many f union [])
-      ; ("and", many f inter [])
-      ; ("not", many f not_union [])
-      ]
-  in
-  let elt =
-    let+ e = f in
-    Element e
-  in
-  peek_exn >>= function
-  | Atom (loc, A "\\") -> User_error.raise ~loc [ Pp.text "unexpected \\" ]
-  | Atom (_, A "") | Quoted_string (_, _) | Template _ -> elt
-  | Atom (loc, A s) -> (
-    match s with
-    | ":standard" -> junk >>> return Standard
-    | ":include" ->
-      User_error.raise ~loc
-        [ Pp.text ":include isn't supported in the predicate language" ]
-    | _ when s.[0] = ':' ->
-      User_error.raise ~loc [ Pp.textf "undefined symbol %s" s ]
-    | _ -> elt)
-  | List (_, Atom (loc, A s) :: _) -> (
-    match s with
-    | ":include" ->
-      User_error.raise ~loc
-        [ Pp.text ":include isn't supported in the predicate language" ]
-    | "or" | "and" | "not" -> bool_ops ()
-    | s when s <> "" && s.[0] <> '-' && s.[0] <> ':' ->
-      User_error.raise ~loc
-        [ Pp.text
-            "This atom must be quoted because it is the first element of a \
-             list and doesn't start with - or:"
+let rec decode_one =
+  let not_union a = compl (union a) in
+
+  fun f ->
+    let open Decoder in
+    let bool_ops () =
+      sum
+        [ ("or", many f union [])
+        ; ("and", many f inter [])
```

Comment:
Looks like we should just call these combinators `or` and `and` instead of `union` and `inter`.

Reply:
Done. Although it's `or_` and `and_` because and/or are keywords.